test: additional approver transfer tests

0xzoz opened the pull request:
## Motivation

This PR address issue #173 and adds 4 new tests testing transfers from the approver

## Change Summary

Tests added:

- testFuzzCannotSafeTransferFromApproverIfFnameExpired
- testFuzzCannotSafeTransferFromApproverIfPaused
- testFuzzCannotTransferFromApproverIfFnameExpired
- testFuzzCannotTransferFromApproverIfPaused

## Merge Checklist

_Choose all relevant options below by adding an `x` now or at any time before submitting for review_

- [x] The PR title adheres to the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) standard
- [x] The PR has been tagged with change type label(s) (i.e. documentation, feature, bugfix, or chore)
- [x] The PR's changes adhere to all the requirements in the [contribution guidelines](https://github.com/farcasterxyz/contracts/blob/main/CONTRIBUTING.md#3-proposing-changes)
- [x] This PR does not require changes to the [Farcaster protocol](https://github.com/farcasterxyz/protocol)

## Additional Context

If this is a relatively large or complex change, provide more details here that will help reviewers.

varunsrin commented on test/NameRegistry.t.sol:
```diff
@@ -1597,6 +1597,59 @@ contract NameRegistryTest is Test {
+        vm.expectRevert(NameRegistry.Expired.selector);
+        nameRegistry.safeTransferFrom(alice, bob, ALICE_TOKEN_ID);
+
+        assertEq(nameRegistry.balanceOf(alice), 1);
```
since this is a secondary test, we can omit these 8 lines, and just do them once at the very end

varunsrin commented on test/NameRegistry.t.sol:
```diff
@@ -1625,6 +1678,45 @@ contract NameRegistryTest is Test {
+        _assumeClean(recovery);
+        _assumeClean(approver);
+        vm.assume(bob != address(0));
+        vm.assume(alice != bob);
```
nit: do all the alice assertions in one line like `vm.assume(alice != bob && alice != approver);`

i know some of the existing tests also have this issue, ive been meaning to clean them up

varunsrin commented on test/NameRegistry.t.sol:
```diff
@@ -1625,6 +1678,45 @@ contract NameRegistryTest is Test {
+
+        uint256 requestTs = _requestRecovery(alice, recovery);
+
+        // alice sets charlie as her approver
```
this comment is incorrect, there is no charlie in this context

varunsrin commented on test/NameRegistry.t.sol:
```diff
@@ -1625,6 +1678,45 @@ contract NameRegistryTest is Test {
+        assertEq(nameRegistry.recoveryDestinationOf(ALICE_TOKEN_ID), recovery);
+        assertEq(nameRegistry.recoveryOf(ALICE_TOKEN_ID), recovery);
+    }
+
```
we should also have a test that shows that a user cannot set an approver if they don't own an fname

varunsrin commented on test/NameRegistry.t.sol:
```diff
@@ -1793,6 +1885,60 @@ contract NameRegistryTest is Test {
+
+        uint256 requestTs = _requestRecovery(alice, recovery);
+
+        // alice sets charlie as her approver
```
incorrect comment

varunsrin commented on test/NameRegistry.t.sol:
```diff
@@ -1793,6 +1885,60 @@ contract NameRegistryTest is Test {
+        vm.expectRevert(NameRegistry.Expired.selector);
+        nameRegistry.transferFrom(alice, bob, ALICE_TOKEN_ID);
+
+        assertEq(nameRegistry.balanceOf(alice), 1);
```
same comment as previously about omitting these assertions

varunsrin commented on test/NameRegistry.t.sol:
```diff
@@ -1822,6 +1968,46 @@ contract NameRegistryTest is Test {
+        _assumeClean(recovery);
+        _assumeClean(approver);
+        vm.assume(bob != address(0));
+        vm.assume(alice != bob);
```
same comment about merging alice assertions into one line

varunsrin commented on test/NameRegistry.t.sol:
```diff
@@ -1822,6 +1968,46 @@ contract NameRegistryTest is Test {
+        assertEq(nameRegistry.ownerOf(ALICE_TOKEN_ID), alice);
+        assertEq(nameRegistry.recoveryTsOf(ALICE_TOKEN_ID), requestTs);
+        assertEq(nameRegistry.recoveryDestinationOf(ALICE_TOKEN_ID), recovery);
+
```
no empty lines in assertion blocks

varunsrin reviewed (changes requested):
Overall looks great, thanks @0xzoz !

Left a few minor comments

0xzoz commented:
Thanks for the pointers @varunsrin  and for looking over this. I have addressed the issues and added the extra test checking only owners of fnames can set approvers.

I also went through and refactored the other `assume` statements in this test based on your feedback.

varunsrin commented on test/NameRegistry.t.sol:
```diff
@@ -1470,12 +1468,40 @@ contract NameRegistryTest is Test {
         nameRegistry.ownerOf(ALICE_TOKEN_ID);
     }
 
+    function testFuzzCannotSetApproverIfNotOwner(
```
this test can be simplified quite a bit - i don't think you need to have recovery or many of the assertions, you can do something like: 


```suggestion
    function testFuzzCannotSetApproverIfNotOwner(
        address alice,
        address bob,
        address approver
    ) public {
        _assumeClean(alice);
        _assumeClean(recovery);
        _assumeClean(approver);
        vm.assume(alice != bob && alice != approver);
        vm.assume(bob != address(0));
        _register(alice);

        // bob tries to set an approver for alices fname
        vm.prank(bob);
        vm.expectRevert("ERC721: approve caller is not token owner or approved for all");
        nameRegistry.approve(approver, ALICE_TOKEN_ID);

       // call ERC721 approver and assert that it is not bob one for alice's token id 
    }

```

0xzoz replied:
Awesome, yeah wasnt sure if I should use these redundant checks. They have been removed. Appreciate the guidance.

varunsrin reviewed (changes requested):
LGTM thanks for cleaning up the other instances as well. I left one comment to simplify the test. 

You can ignore the coverage test failure, that's something I introduced and will fix separately
